---
title: "Aligning reads with Rhisat2"
author: "Charlotte Soneson"
package: Rhisat2
output: 
    BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{Rhisat2}
    %\VignetteEncoding{UTF-8}  
    %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
options(width=100)
```

The `r Biocpkg("Rhisat2")` package provides an interface to the
[hisat2](https://ccb.jhu.edu/software/hisat2/index.shtml) short-read aligner
software.

```{r}
library(Rhisat2)
```

# Building a genome index

To build an index for the alignment, use the `hisat2_build` function. You need
to provide one or more fasta file with reference sequences, as well as an output
directory where the index will be stored, and a "prefix" (that will determine
the name of the index files in the output directory).

The package contains three example reference sequences, corresponding to short
pieces of three different chromosomes. We show how to create an index (named
`myindex`) based on these sequences.

```{r}
list.files(system.file("extdata/refs", package="Rhisat2"))
refs <- list.files(system.file("extdata/refs", package="Rhisat2"), 
                   full.names=TRUE)
td <- tempdir()
hisat2_build(references=refs, outdir=td, prefix="myindex", 
             force=TRUE, strict=TRUE, execute=TRUE)
```

By setting `execute=FALSE`, `hisat2_build()` will return the index building shell
command, without executing it.

```{r}
print(hisat2_build(references=refs, outdir=td, prefix="myindex",
                   force=TRUE, strict=TRUE, execute=FALSE))
```

# Aligning reads to the genome index

After creating the index, reads can be aligned using the `hisat2` function. Most
commonly, the reads will be provided in `fastq` files (one for single-end reads,
two for paired-end reads). The names of these files can be provided directly to
the `hisat2` function, either as a vector (for single-end reads) or as a list of
length 2 (for paired-end reads, each list element corresponds to one mate). You
also need to provide the path to the index generated by `hisat2_build` (the
output directory combined with the prefix), and the output file name where the
alignments should be written.

```{r}
list.files(system.file("extdata/reads", package="Rhisat2"))
reads <- list.files(system.file("extdata/reads", package="Rhisat2"),
                    pattern="\\.fastq$", full.names=TRUE)
hisat2(sequences=as.list(reads), index=file.path(td, "myindex"), 
       type="paired", outfile=file.path(td, "out.sam"), 
       force=TRUE, strict=TRUE, execute=TRUE)
```

# Miscellaneous helper functions

To get the version of `hisat2`:

```{r}
hisat2_version()
```

To see the usage of `hisat2_build()`:

```{r}
hisat2_build_usage()
```

And similarly to see the usage of `hisat2()`:

```{r}
hisat2_usage()
```

# Session info

```{r}
sessionInfo()
```

